azuread/data_users.go

```diff
@@ -62,54 +126,83 @@ func dataSourceUsersRead(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*ArmClient).usersClient
 	ctx := meta.(*ArmClient).StopContext
 
-	var users []graphrbac.User
+	var users []*graphrbac.User
 	expectedCount := 0
 
+	gracefulError := d.Get("graceful_errors").(bool)
 	if upns, ok := d.Get("user_principal_names").([]interface{}); ok && len(upns) > 0 {
 		expectedCount = len(upns)
 		for _, v := range upns {
-			resp, err := client.Get(ctx, v.(string))
+			u, err := client.Get(ctx, v.(string))
 			if err != nil {
-				return fmt.Errorf("Error making Read request on AzureAD User with ID %q: %+v", v.(string), err)
+				if !gracefulError {
+					return fmt.Errorf("Error making Read request on AzureAD User with ID %q: %+v", v.(string), err)
+				}
+				break
 			}
-
-			users = append(users, resp)
+			users = append(users, &u)
 		}
 	} else if oids, ok := d.Get("object_ids").([]interface{}); ok && len(oids) > 0 {
 		expectedCount = len(oids)
 		for _, v := range oids {
 			u, err := graph.UserGetByObjectId(&client, ctx, v.(string))
 			if err != nil {
-				return fmt.Errorf("Error finding Azure AD User with object ID %q: %+v", v.(string), err)
+				if !gracefulError {
+					return fmt.Errorf("Error finding Azure AD User with object ID %q: %+v", v.(string), err)
+				}
+				break
 			}
-			users = append(users, *u)
+			users = append(users, u)
 		}
 	} else if mailNicknames, ok := d.Get("mail_nicknames").([]interface{}); ok && len(mailNicknames) > 0 {
 		expectedCount = len(mailNicknames)
 		for _, v := range mailNicknames {
 			u, err := graph.UserGetByMailNickname(&client, ctx, v.(string))
 			if err != nil {
-				return fmt.Errorf("Error finding Azure AD User with email alias %q: %+v", v.(string), err)
+				if !gracefulError {
+					return fmt.Errorf("Error finding Azure AD User with email alias %q: %+v", v.(string), err)
+				}
+				break
 			}
-			users = append(users, *u)
+			users = append(users, u)
 		}
 	}
 
-	if len(users) != expectedCount {
+	if !gracefulError && len(users) != expectedCount {
 		return fmt.Errorf("Unexpected number of users returned (%d != %d)", len(users), expectedCount)
 	}
 
+	if gracefulError && len(users) == 0 {
+		return fmt.Errorf("No users were returned")
+	}
+
 	upns := make([]string, 0, len(users))
 	oids := make([]string, 0, len(users))
 	mailNicknames := make([]string, 0, len(users))
+	var userList []map[string]interface{}
 	for _, u := range users {
-		if u.ObjectID == nil || u.UserPrincipalName == nil {
+		if !gracefulError && (u.ObjectID == nil || u.UserPrincipalName == nil || u.MailNickname == nil) {
```

Comment:
Did you find it necessary to skip this error condition when testing? I'm thinking we may want to err here in every case since it indicates a bad API response. Also I think `MailNickname` might be optional for some accounts so it being nil could be ok?

Reply:
Ah, didn't realise that about `MailNickname`. Yes this can be reverted.